rclpy/rclpy/executors.py

```diff
@@ -222,11 +227,15 @@ def _take_client(self, client):
 
     async def _execute_client(self, client, seq_and_response):
         sequence, response = seq_and_response
-        if sequence is not None and sequence == client.sequence_number:
-            # clients spawn their own thread to wait for a response in the
-            # wait_for_future function. Users can either use this mechanism or monitor
-            # the content of client.response to check if a response has been received
-            client.response = response
+        if sequence is not None:
+            try:
+                future = client._pending_requests[sequence]
+            except IndexError:
+                # The request was cancelled
+                pass
+            else:
+                future._set_executor(self)
```

Comment:
It seems weird that the executor is calling a "private" method on the future?

Reply:
Ideas? It seems weird to me too. The future needs a reference to an executor to be able to add callbacks.

`asyncio` solves the problem by expecting futures to be created via `executor.create_future()` which [passes itself as a keyword argument](https://github.com/python/cpython/blob/bec2372b7e1da5dfdbadaf242aa8e994b164cace/Lib/asyncio/base_events.py#L373-L375) to `Future.__init__()`.

Right now the client doesn't have a reference to the executor, so rclpy can't do the same thing. Instead the client holds onto the future and hopes the user will add the node to an executor.

Reply:
I don't have a suggestion as to how to change it, and I guess @dirk-thomas doesn't either?

So, I think it's ok to merge like this.